src/payment/strategies/square/square-payment-strategy.ts

```diff
@@ -43,13 +66,29 @@ export default class SquarePaymentStrategy extends PaymentStrategy {
     }
 
     execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
-        const { payment, ...order } = payload;
+        const { payment } = payload;
 
         if (!payment || !payment.methodId) {
-            throw new InvalidArgumentError('Unable to submit payment because "payload.payment.methodId" argument is not provided.');
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
         }
 
-        const paymentName = payment.methodId;
+        const { methodId } = payment;
+
+        return this._getPaymentData(payment)
+            .then(paymentData => {
+                const paymentPayload = { methodId, paymentData };
+
+                return this._store.dispatch(this._orderActionCreator.submitOrder(payload, options))
+                    .then(() =>
+                        this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload))
+                    );
+            });
+    }
+
+    private _getPaymentData({ paymentData }: OrderPaymentRequestBody): Promise<NonceInstrument> {
+        if (paymentData && isNonceLike(paymentData)) {
```

Comment:
I just had a closer look at this. You know you shouldn't pass in the `nonce` from the outside like this. You see, for the non-wallet flow, how we retrieve the nonce and forward it to `submitPayment` within the payment strategy? You need to do the same thing for Masterpass - because third party developers wouldn't know that they have to grab some undocumented property from `initializationData` and pass it to `CheckoutService#submitOrder`.

Reply:
👍